ps.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"text/tabwriter"
+
+	"github.com/hyperhq/runv/containerd/api/grpc/types"
+	"github.com/urfave/cli"
+	netcontext "golang.org/x/net/context"
+)
+
+var psCommand = cli.Command{
+	Name:      "ps",
+	Usage:     "ps displays the processes running inside a container",
+	ArgsUsage: `<container-id> [ps options]`, Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "format, f",
+			Value: "table",
+			Usage: `select one of: ` + formatOptions,
+		},
+	},
+	Action: func(context *cli.Context) error {
+		container := context.Args().First()
+		if container == "" {
+			return cli.NewExitError("container id cannot be empty", -1)
+		}
+		c, err := getContainerApi(context, container)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("can't access container, %v", err), -1)
+		}
+
+		switch context.String("format") {
+		case "table":
+			w := tabwriter.NewWriter(os.Stdout, 12, 1, 3, ' ', 0)
+			fmt.Fprint(w, "PROCESS\tCMD\n")
+			// we are limited by the containerd interface for now
+			for _, p := range c.Processes {
+				fmt.Fprintf(w, "%s\t%s\n",
+					p.Pid,
+					p.Args)
+			}
+			if err := w.Flush(); err != nil {
+				fatal(err)
+			}
+		case "json":
+			pids := make([]string, 0)
+			for _, p := range c.Processes {
+				pids = append(pids, p.Pid)
+			}
+
+			data, err := json.Marshal(pids)
+			if err != nil {
+				fatal(err)
+			}
+			os.Stdout.Write(data)
```

Comment:
SystemPid reports the PID of the qemu process on the host for the container's main process, and AFAIU is invalid for the rest. See my original comment on this topic, in the first comment of this thread

> Runv currently implements the OCI interface by wrapping runv-containerd calls, so we are limited by the containerd interface as supported by runv at the moment.
> This introduces some limitations on these implementations, which to resolve would require the refactoring or runv, which IMHO should be discussed.
> 
> First, the processes runv is aware of from each container is limited to only the processes started directly from runv (via run and exec). Children are not being tracked. An extension in hyperstart would be needed to overcome this limitation (to always fetch the process list dynamically).
> 
> Additionally, runv-containerd reports only limited information about the processes running inside the container. We can not show the PID for example. I'm not sure whether extending hyperstart to provide this info is enough, as we need to consider how it will affect runv-containerd. Even then, we will be limited by the current fields in the containerd API. In any case runv-containerd and the runv OCI interface need to be considered at the same time to improve upon this point.